packages/next/server/web/spec-extension/response.ts

```diff
@@ -71,15 +90,22 @@ export class NextResponse extends Response {
     })
   }
 
-  static rewrite(destination: string | NextURL | URL, init?: ResponseInit) {
+  static rewrite(
+    destination: string | NextURL | URL,
+    init?: MiddlewareResponseInit
+  ) {
     const headers = new Headers(init?.headers)
     headers.set('x-middleware-rewrite', validateURL(destination))
+
+    handleMiddlewareField(init, headers)
```

Comment:
For `json()`, yes. We don't allow returning a response in Next.js middleware so adding this feature to `json()` is meaningless and confusing.

Secondly, we (web servers) can't specify request headers in redirections (`redirect`) because browsers make another request, while `next()` and `rewrite()` are handled internally in server-side.

Reply:
> We don't allow returning a response in Next.js middleware

I think it'd be good to ensure the API surface is consistent regardless of this limitation x-ref: [slack thread](https://vercel.slack.com/archives/C0289CGVAR2/p1665451610968199?thread_ts=1665169221.188349&cid=C0289CGVAR2)

Reply:
@ijjk Wow didn't know #41243. However, still not sure if we should add this API in `NextResponse.json()`. Even if the middleware can return a response, we skip the actual data fetch from Serverless/Edge Functions, right? If so it doesn't make sense to allow mutating *request* headers. Btw I don't have a strong opinion. Just want to clarify 😄

Reply:
Ah yeah if we're shortcutting the serverless/edge function it make sense to not allow specifying this here